File: backend/workline/knowledge/cache/cache.py
"""Central KnowledgeCache facade coordinating L1 Memory, L2 Persistent storage, and invalidation."""

import json
import threading
import time
from typing import Any, Dict, List, Optional
from backend.workline.knowledge.cache.invalidation import InvalidationManager
from backend.workline.knowledge.cache.memory import MemoryCache
from backend.workline.knowledge.cache.models import (
    CacheMetadata,
    CacheObjectType,
    CacheOptions,
    CacheStats,
)
from backend.workline.knowledge.cache.persistent import PersistentCache


class KnowledgeCache:
    """
    Two-tiered cache engine (L1 in-process memory + L2 persistent filesystem).
    Ensures non-authoritative caching, strict TTL, and project/team isolation.
    """

    def __init__(
        self,
        l1_max_entries: int = 2000,
        l2_base_dir: str = ".workline/cache",
    ):
        self.l1 = MemoryCache(l1_max_entries)
        self.l2 = PersistentCache(l2_base_dir)
        self.invalidation = InvalidationManager()
        self._lock = threading.RLock()
        self._stats = CacheStats()

    def get(self, key: str, object_type: CacheObjectType = CacheObjectType.RETRIEVAL) -> Optional[Any]:
        """Fetch item from L1 memory or fallback to L2 persistent cache."""
        with self._lock:
            # 1. Check L1 Memory
            l1_item = self.l1.get(key)
            if l1_item is not None:
                self._stats.hits += 1
                return l1_item[1]

            # 2. Check L2 Persistent
            l2_item = self.l2.get(key, object_type)
            if l2_item is not None:
                meta, data = l2_item
                # Promote to L1
                self.l1.set(key, meta, data)
                self._stats.hits += 1
                return data

            self._stats.misses += 1
            return None

    def set(
        self,
        key: str,
        value: Any,
        object_type: CacheObjectType,
        options: CacheOptions,
    ) -> None:
        """Store item into both L1 and L2 caches."""
        now = time.time()
        ttl_seconds = options.ttl if options.ttl is not None else self._get_default_ttl(object_type)
        expires_at = now + ttl_seconds if ttl_seconds > 0 else 0.0

        try:
            size_bytes = len(json.dumps(value, default=str))
        except Exception:
            size_bytes = 100

        meta = CacheMetadata(
            cache_key=key,
            object_type=object_type,
            project_id=options.project_id,
            team_id=options.team_id,
            source_id=options.source_id,
            source_hash=options.source_hash,
            schema_version=options.schema_version,
            created_at=now,
            expires_at=expires_at,
            project_version=options.project_version,
            git_commit=options.git_commit,
            provider=options.provider,
            model=options.model,
            embedding_dimension=options.embedding_dimension,
            size_bytes=size_bytes,
        )

        with self._lock:
            self.l1.set(key, meta, value)
            self.l2.set(key, meta, value)
            self.invalidation.track(key, options.project_id, options.source_id)

    def has(self, key: str, object_type: CacheObjectType = CacheObjectType.RETRIEVAL) -> bool:
        return self.get(key, object_type) is not None

    def delete(self, key: str, object_type: CacheObjectType = CacheObjectType.RETRIEVAL) -> bool:
        with self._lock:
            self.l1.delete(key)
            deleted_l2 = self.l2.delete(key, object_type)
            self.invalidation.remove_key(key)
            return deleted_l2

    def invalidate_by_source(self, source_id: str) -> int:
        with self._lock:
            keys = self.invalidation.get_keys_by_source(source_id)
            for k in keys:
                self.l1.delete(k)
                for t in CacheObjectType:
                    self.l2.delete(k, t)
            self._stats.invalidations += len(keys)
            return len(keys)

    def invalidate_by_project(self, project_id: str) -> int:
        with self._lock:
            keys = self.invalidation.get_keys_by_project(project_id)
            for k in keys:
                self.l1.delete(k)
                for t in CacheObjectType:
                    self.l2.delete(k, t)
            self._stats.invalidations += len(keys)
            return len(keys)

    def clear(self) -> None:
        with self._lock:
            self.l1.clear()
            self.l2.clear()
            self.invalidation.clear()

    def clear_expired(self) -> int:
        with self._lock:
            expired_l1 = self.l1.clear_expired()
            self._stats.expired += expired_l1
            return expired_l1

    def get_stats(self) -> CacheStats:
        with self._lock:
            l2_entries, l2_size = self.l2.get_stats()
            self._stats.l1_entries = self.l1.size()
            self._stats.l2_entries = l2_entries
            self._stats.l2_size_bytes = l2_size

            total = self._stats.hits + self._stats.misses
            self._stats.hit_rate = (self._stats.hits / total * 100.0) if total > 0 else 0.0
            self._stats.miss_rate = (self._stats.misses / total * 100.0) if total > 0 else 0.0

            return self._stats.model_copy()

    def _get_default_ttl(self, object_type: CacheObjectType) -> int:
        if object_type in (CacheObjectType.DOCUMENT_PARSE, CacheObjectType.DOCUMENT_CHUNK, CacheObjectType.EMBEDDING):
            return 86400 * 7  # 7 days
        if object_type in (CacheObjectType.SUMMARY, CacheObjectType.RESEARCH):
            return 3600 * 24  # 24 hours
        if object_type == CacheObjectType.RETRIEVAL:
            return 3600 * 2   # 2 hours
        return 300            # 5 minutes for Context and Discovery


# Global singleton instance
knowledge_cache = KnowledgeCache()

File: packages/knowledge/cache/cache.ts
/**
 * Unified KnowledgeCache coordinating L1 memory and L2 persistent storage.
 */

import {
  CacheEntry,
  CacheMetadata,
  CacheObjectType,
  CacheOptions,
  CacheStats,
} from "./metadata";
import { MemoryCache } from "./memory";
import { PersistentCache } from "./persistent";
import { InvalidationEngine } from "./invalidation";
import { CURRENT_SCHEMA_VERSION } from "./serializer";

export class KnowledgeCache {
  private l1: MemoryCache;
  private l2: PersistentCache;
  private invalidation: InvalidationEngine;
  private stats: CacheStats;

  constructor(
    l1MaxEntries: number = 2000,
    l2BaseDir: string = ".workline/cache"
  ) {
    this.l1 = new MemoryCache(l1MaxEntries);
    this.l2 = new PersistentCache(l2BaseDir);
    this.invalidation = new InvalidationEngine();
    this.stats = {
      hits: 0,
      misses: 0,
      evictions: 0,
      expired: 0,
      invalidations: 0,
      l1Entries: 0,
      l2Entries: 0,
      l2SizeBytes: 0,
      hitRate: 0.0,
      missRate: 0.0,
    };
  }

  public async get<T>(key: string, objectType: CacheObjectType = CacheObjectType.RETRIEVAL): Promise<T | null> {
    // 1. Try L1 Memory
    const l1Entry = this.l1.get<T>(key);
    if (l1Entry) {
      this.recordHit();
      return l1Entry.data;
    }

    // 2. Try L2 Persistent
    const l2Entry = this.l2.get<T>(key, objectType);
    if (l2Entry) {
      // Promote to L1
      this.l1.set(l2Entry);
      this.recordHit();
      return l2Entry.data;
    }

    this.recordMiss();
    return null;
  }

  public async set<T>(
    key: string,
    value: T,
    objectType: CacheObjectType,
    options: CacheOptions
  ): Promise<void> {
    const now = Date.now();
    const ttlSeconds = options.ttl ?? this.getDefaultTtl(objectType);
    const expiresAt = ttlSeconds > 0 ? now + ttlSeconds * 1000 : 0;

    const metadata: CacheMetadata = {
      cacheKey: key,
      objectType,
      projectId: options.projectId,
      teamId: options.teamId ?? "default_team",
      sourceId: options.sourceId,
      sourceHash: options.sourceHash,
      schemaVersion: options.schemaVersion ?? CURRENT_SCHEMA_VERSION,
      createdAt: now,
      expiresAt,
      projectVersion: options.projectVersion,
      gitCommit: options.gitCommit,
      provider: options.provider,
      model: options.model,
      embeddingDimension: options.embeddingDimension,
      sizeBytes: JSON.stringify(value).length,
    };

    const entry: CacheEntry<T> = { metadata, data: value };

    // Write to L1 and L2
    this.l1.set(entry);
    this.l2.set(entry);
    this.invalidation.track(entry);
  }

  public async has(key: string, objectType: CacheObjectType = CacheObjectType.RETRIEVAL): Promise<boolean> {
    return (await this.get(key, objectType)) !== null;
  }

  public async delete(key: string, objectType: CacheObjectType = CacheObjectType.RETRIEVAL): Promise<boolean> {
    this.l1.delete(key);
    const deletedL2 = this.l2.delete(key, objectType);
    this.invalidation.removeKey(key);
    return deletedL2;
  }

  public async invalidateBySource(sourceId: string): Promise<number> {
    const keys = this.invalidation.getKeysBySource(sourceId);
    for (const k of keys) {
      this.l1.delete(k);
      for (const t of Object.values(CacheObjectType)) {
        this.l2.delete(k, t as CacheObjectType);
      }
    }
    this.stats.invalidations += keys.length;
    return keys.length;
  }

  public async invalidateByProject(projectId: string): Promise<number> {
    const keys = this.invalidation.getKeysByProject(projectId);
    for (const k of keys) {
      this.l1.delete(k);
      for (const t of Object.values(CacheObjectType)) {
        this.l2.delete(k, t as CacheObjectType);
      }
    }
    this.stats.invalidations += keys.length;
    return keys.length;
  }

  public async clear(): Promise<void> {
    this.l1.clear();
    this.l2.clear();
    this.invalidation.clear();
  }

  public async clearExpired(): Promise<number> {
    const l1Expired = this.l1.clearExpired();
    this.stats.expired += l1Expired;
    return l1Expired;
  }

  public getStats(): CacheStats {
    const l2Stats = this.l2.getStats();
    this.stats.l1Entries = this.l1.size();
    this.stats.l2Entries = l2Stats.totalEntries;
    this.stats.l2SizeBytes = l2Stats.totalSizeBytes;

    const totalReqs = this.stats.hits + this.stats.misses;
    this.stats.hitRate = totalReqs > 0 ? (this.stats.hits / totalReqs) * 100 : 0.0;
    this.stats.missRate = totalReqs > 0 ? (this.stats.misses / totalReqs) * 100 : 0.0;

    return { ...this.stats };
  }

  private getDefaultTtl(objectType: CacheObjectType): number {
    switch (objectType) {
      case CacheObjectType.DOCUMENT_PARSE:
      case CacheObjectType.DOCUMENT_CHUNK:
      case CacheObjectType.EMBEDDING:
        return 86400 * 7; // 7 days
      case CacheObjectType.SUMMARY:
      case CacheObjectType.RESEARCH:
        return 3600 * 24; // 24 hours
      case CacheObjectType.RETRIEVAL:
        return 3600 * 2;  // 2 hours
      case CacheObjectType.CONTEXT:
      case CacheObjectType.AGENT_DISCOVERY:
      default:
        return 300;       // 5 minutes
    }
  }

  private recordHit(): void {
    this.stats.hits++;
  }

  private recordMiss(): void {
    this.stats.misses++;
  }
}

// Global singleton instance
export const knowledgeCache = new KnowledgeCache();
